pybamm/discretisations/discretisation.py

```diff
@@ -1197,3 +1200,85 @@ def check_variables(self, model):
+                    pybamm.tree_search(rhs_child, var, this_var_list)
+        for tree in rhs_variables:
+            pybamm.tree_search(model.rhs[tree], var, this_var_list)
+        for tree in algebraic_variables:
+            pybamm.tree_search(model.algebraic[tree], var, this_var_list)
+        for (keys, tree) in zip(boundary_variable_keys, boundary_variables):
+            for key in keys:
+                pybamm.tree_search(
+                    model.boundary_conditions[tree][key][0], var, this_var_list
+                )
+        for tree in model.variables.keys():
+            for rhs_child in model.variables[tree].children:
+                pybamm.tree_search(rhs_child, var, this_var_list)
+        this_var_is_independent = not any(this_var_list)
+        not_in_y_slices = not (var in list(self.y_slices.keys()))
+        not_in_discretised = not (var in list(self._discretised_symbols.keys()))
+        is_0D = len(var.domain) == 0
+        this_var_is_independent = (
+            this_var_is_independent and not_in_y_slices and not_in_discretised and is_0D
+        )
+        return model, this_var_is_independent
+
+    def check_for_independent_variables(self, model):
+        rhs_vars_to_search_over = list(model.rhs.keys())
+        for var in rhs_vars_to_search_over:
+            model, this_var_is_independent = self.search_for_independent_var(model, var)
+            if this_var_is_independent:
+                if len(model.rhs) != 1:
+                    pybamm.logger.info("removing variable {} from rhs".format(var))
+                    my_initial_condition = model.initial_conditions[var]
+                    model.variables[var.name] = pybamm.ExplicitTimeIntegral(
+                        model.rhs[var], my_initial_condition
+                    )
+                    # edge case where a variable appears
+                    # in the rhs twice under different names
+                    for key in model.variables:
+                        if model.variables[key] == var:
+                            model.variables[key] = model.variables[var.name]
+                    del model.rhs[var]
+                    del model.initial_conditions[var]
+                else:
+                    break
+        algebraic_vars_to_search_over = list(model.algebraic.keys())
+        for var in algebraic_vars_to_search_over:
+            model, this_var_is_independent = self.search_for_independent_var(model, var)
+            if this_var_is_independent:
+                if len(model.algebraic) != 1:
+                    pybamm.logger.info(
+                        "removing variable {} from algebraic".format(var)
+                    )
+                    my_initial_condition = model.initial_conditions[var]
+                    model.variables[var.name] = model.algebraic[var]
+                    del model.algebraic[var]
+                    del model.initial_conditions[var]
+                else:
+                    break
```

Comment:
We shouldn't do this for algebraic variables. The algebraic dictionary is `0=...`, *not* `key=value` like it looks like you are assuming here. For example, `0=div(sigma*grad(phi_s)) + a*j` for Ohm's law. The key in the dictionary defines which variable that equation "solves" for, not equality between the variable and the equation.